### PR TITLE
[mlir][vector] Add deinterleave operation to vector dialect

### PR DESCRIPTION
**Pull request by mub-at-arm**
The deinterleave operation constructs two vectors from a single input
vector. The first result vector contains the elements from even indexes
of the input, and the second contains elements from odd indexes. This is
the inverse of a `vector.interleave` operation.

Each output's trailing dimension is half of the size of the input
vector's trailing dimension. This operation requires the input vector
to have a rank > 0 and an even number of elements in its trailing
dimension.

The operation supports scalable vectors.

Example:
```mlir
%0, %1 = vector.deinterleave %a
           : vector<8xi8> -> vector<4xi8>
%2, %3 = vector.deinterleave %b
           : vector<2x8xi8> -> vector<2x4xi8>
%4, %5 = vector.deinterleave %c
           : vector<2x8x4xi8> -> vector<2x8x2xi8>
%6, %7 = vector.deinterleave %d
           : vector<[8]xf32> -> vector<[4]xf32>
%8, %9 = vector.deinterleave %e
           : vector<2x[6]xf64> -> vector<2x[3]xf64>
%10, %11 = vector.deinterleave %f
           : vector<2x4x[6]xf64> -> vector<2x4x[3]xf64>
```

### CODE REVIEW

File: mlir/test/Target/LLVMIR/llvmir-intrinsics.mlir
Comment (MacDue) [inline mlir/test/Target/LLVMIR/llvmir-intrinsics.mlir:953]: It looks like you need to rebase this PR, looks like it contains some previous changes.

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:579]: nit: I'd drop "essentially" (this is the inverse).
```suggestion
        from the input, respectively. This is the inverse of an
```

Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:579]: I think this could be clear:
```suggestion
        vector. The first result vector contains the elements from even 
        lanes of the input, and the second contains elements from odd lanes.
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:585]: Maybe reword a little?:
```suggestion
        The trailing dimension of each output is half of the size of input vector's 
        trailing dimension. This operation requires that the input vector has a 
        rank > 0 and a even number of elements its trailing dimension.
```

**Review by MacDue — COMMENTED** [review]
A few comments on the documentation:

**Comment by MacDue** [comment]
Marked as draft due to the rebase issue. I'll help @mub-at-arm resolve the issues with this PR tomorrow.

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:602]: nit: Rename `$input` to `$source` (and update the getters/references accordingly), that'd be more consistent with other operations in the vector dialect. 
```suggestion
      let arguments = (ins AnyVector:$source);
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:547]: nit: specify this checks the trailing dimension?
```suggestion
  "the trailing dimension of the results is half the width of source trailing dimension",
```

File: mlir/test/Dialect/Vector/invalid.mlir
Comment (MacDue) [inline mlir/test/Dialect/Vector/invalid.mlir:1856]: very minor nit: Can you ensure your editor adds newlines at the end of files (and update two test files)?
See [here](https://stackoverflow.com/questions/44704968/visual-studio-code-insert-newline-at-the-end-of-files) for a guide on this for vscode.

**Review by MacDue — COMMENTED** [review]
a few more nits:

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:598]: The docs should show this returns two results (I don't think assigning it to one value is valid): 
```suggestion
        %0, %1 = vector.deinterleave %a
                   : vector<[4]xi32>     ; yields vector<[2]xi32>, vector<[2]xi32>
        %2, %3 = vector.deinterleave %b
                   : vector<8xi8>        ; yields vector<4xi8>, vector<4xi8>
        %4, %5 = vector.deinterleave %c
                   : vector<2x8xf32>     ; yields vector<2x4xf32>, vector<2x4xf32>
        %6, %7 = vector.deinterleave %d
                   : vector<2x4x[6]xf64> ; yields vector<2x4x[3]xf64>, vector<2x4x[3]xf64>
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:592]: What do you think about adding the result type? I think this will make it more readable but not too verbose, and we won't have to add comments like this
```mlir
%0, %1 = vector.deinterleave %a : vector<[4]xi32> into vector<[2]xi32>
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:618]: Aren't these the same?

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (MacDue) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:580]: 'lane' in the `vector.mask` documentation refers to the same concept in interleave/deinterleave.

Comment (MacDue) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:580]: but you could say "elements from even indexes..." instead of lanes

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (MacDue) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:618]: They are -- there should be a single  `getResultVectorType()` method +1

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:610]: The examples and `assemblyFormat` don't match. Also, I think @kuhar's suggestion was just to show one of the result types (since the other is the same). 

e.g.:
```
$source attr-dict `:` type($source) `into` type($res1)
```

Comment (mub-at-arm) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: @MacDue, @kuhar, I've added and `assemblyFormat` that fully describes the operation without being too verbose. An "x2" suffix on the result describes the quantity of that type quite simply and should be easily readable too.

So the resulting IR should look like this:
`%1, %2 = vector.deinterleave %a : vector<4xf32> -> vector<2xf32>x2`

Comment (MacDue) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: That looks good to me :+1:

**Review by MacDue — APPROVED** [review]
This LGTM, though I'll wait to see if @dcaballe or @nicolasvasilache have any comments :slightly_smiling_face:

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (c-rhodes) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: I'm not a fan of that assembly format, I think consistency is much more important and would rather have the verbose form.

Comment (MacDue) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: The verbose form is not consistent with `vector.interleave` (which only shows the source type) -- (similar to `vectpr.shuffle`)

Comment (c-rhodes) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: I meant consistency in general not in relation to interleave.

Comment (mub-at-arm) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: Should it be consistent to interleave or the other MLIR operation formats?

Comment (kuhar) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: > @MacDue, @kuhar, I've added and `assemblyFormat` that fully describes the operation without being too verbose. An "x2" suffix on the result describes the quantity of that type quite simply and should be easily readable too.
> 
> So the resulting IR should look like this: `%1, %2 = vector.deinterleave %a : vector<4xf32> -> vector<2xf32>x2`

I'd prefer just `%1, %2 = vector.deinterleave %a : vector<4xf32> -> vector<2xf32>` for readability and consistency with most of core mlir dialects.
Separately, I think we should also have the result type on `vector.interleave` (I wasn't pinged on the initial PR IIRC, and I didn't catch it before it landed).

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:595]: nit: I think it would be more digestible for the readers to lead with easier examples, i.e., simple static 1d vector, and move n-d vectors and scalable vectors towards the bottom.

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:609]: I don't think we need the `x2`, see: https://github.com/llvm/llvm-project/pull/92409#discussion_r1608455010

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:583]:
```suggestion
        the inverse of a `vector.interleave` operation.
```

**Review by kuhar — COMMENTED** [review]
A few cosmetic issues, otherwise looks good

**Review by kuhar — APPROVED** [review]

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:569]: nit: fix indentation

```suggestion
  CPred<[{
    [&](){
      auto srcVec = getSourceVectorType();
      return srcVec.getDimSize(srcVec.getRank() - 1) % 2 == 0;
    }()
  }]>
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:595]: nit: space
```suggestion
                   : vector<8xi8> -> vector<4xi8>
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:605]: nit: update var names
```suggestion
        %4, %5 = vector.deinterleave %c
                   : vector<2x8x4xi8> -> vector<2x8x2xi8>
        %6, %7 = vector.deinterleave %d
                   : vector<[8]xf32> -> vector<[4]xf32>
        %8, %9 = vector.deinterleave %e
                   : vector<2x[6]xf64> -> vector<2x[3]xf64>
        %10, %11 = vector.deinterleave %f
                   : vector<2x4x[6]xf64> -> vector<2x4x[3]xf64>
```

**Review by c-rhodes — APPROVED** [review]
Few minor nits but otherwise LGTM, cheers!

**Comment by MacDue** [comment]
If there's no further comments I'll land this tomorrow :slightly_smiling_face:

**Review by banach-space — APPROVED** [review]

**Pull request merged**
